[webkitbugspy] Add ability to set see_also field for bugs

briannafan opened the pull request:
#### 675b03ad0fde8ec9116b51a893f8766e3cf84871
<pre>
[webkitbugspy] Add ability to set see_also field for bugs
<a href="https://rdar.apple.com/151183507">rdar://151183507</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=292893">https://bugs.webkit.org/show_bug.cgi?id=292893</a>

Reviewed by Jonathan Bedard and Sam Sneddon.

Set the &apos;see_also&apos; field in bugzilla by calling Issue.add_related_link.
This should be passed in as a list of string URLs.

&apos;see_also&apos; is stored in Issue.related_links as a new property since links that are not currently
tracked by webkitbugspy (GitHub, etc.) would not be populated in Issue.references.

* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.populate):
(Tracker.set):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py:
(Issue.__init__):
(Issue):
(Issue.related_links):
(Issue.add_related_links):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py:
(Bugzilla._issue):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker.populate):
(Tracker.set):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/github_unittest.py:
(test_source_changes):
(test_related_links):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:

Canonical link: <a href="https://commits.webkit.org/295169@main">https://commits.webkit.org/295169@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e98ecc6dddb16f0ed4c5f784a6a7371c8043133

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104299 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/24003 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/14416 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/109500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/54967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/106339 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/24375 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/32550 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/109500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/54967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/107305 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/24375 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/14416 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/109500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/103777 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/24375 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/14416 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/54327 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/24375 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/14416 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/111881 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/31456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/32550 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/111881 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/31820 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/14416 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/111881 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/14416 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/25924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16933 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/31384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/31178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/34514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/32738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -401,6 +401,9 @@ def set(self, issue, assignee=None, opened=None, why=None, project=None, compone
 
+        if see_also is not None:
+            update_dict['see_also'] = dict(add=[see_also])
+
```
How come you're wrapping this in a list?

I would expect behaviour to be the same as `keywords` above, which also uses `set` and not `add`.

briannafan replied:
There's actually no `set` for `see_also` ([docs](https://bugzilla.readthedocs.io/en/5.0.4/api/core/v1/bug.html#update-bug)). The API takes a list, so either we need to pass in a list or turn a single element into a list. Maybe it would make more sense to pass in a list in the first place  to follow our convention in other functions.

gsnedders replied:
> There's actually no set for see_also ([docs](https://bugzilla.readthedocs.io/en/5.0.4/api/core/v1/bug.html#update-bug)).

That's weird and whacky. I don't know what the right approach is here, because I'd rather our API be consistent at least?

> The API takes a list, so either we need to pass in a list or turn a single element into a list. Maybe it would make more sense to pass in a list in the first place to follow our convention in other functions.

I'd suggest we pass in a sequence in the first place, because we might want to add multiple?

briannafan replied:
Yep! I've made that change in the current version

gsnedders commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
```diff
@@ -434,6 +434,10 @@ def set(self, issue, assignee=None, opened=None, why=None, project=None, compone
             return None
 
+        if see_also:
+            sys.stderr.write('GitHub does not support the see_also field at this time\n')
```
I'll note that I don't like the fact we just write to `sys.stderr` and then return `None` in what's kinda meant to be a library, but it is nothing but following the status quo.

gsnedders commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/data.py:
```diff
@@ -70,6 +70,7 @@
                 content='Trying to prepare a fix for 1.2.4',
             ),
         ], watchers=[USERS['Felix Filer'], USERS['Tim Contributor'], USERS['Wilma Watcher']],
+        references=[]
```
Is this not part of what forms the `data` in `Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py`? Because we're looking at `data['see_also']['add']` there.

briannafan replied:
Might be more correct to delete this and change the mocks instead.

briannafan replied:
Also, not sure if I'm misunderstanding your comment, but the `data` you're referencing there is not the data from this file but rather from the formed request in `Bugzilla.set`

gsnedders replied:
Ah, OK, so I just don't understand what's going on here.

gsnedders commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py:
```diff
@@ -166,6 +166,8 @@ def _issue(self, url, id, credentials=None, data=None):
             if data.get('regressions'):
                 issue['related']['regressions'] = data['regressions']
+            if data.get('see_also'):
+                issue['references'] += data['see_also']['add']
```
As above, I would've expected similar behaviour to `keywords` (now below).

briannafan replied:
As above (:P), there's no `set` for `see_all`. Then, the logic is essentially the same since we don't need to condition for different API behaviour.

gsnedders commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
```diff
@@ -437,6 +437,10 @@ def set(self, issue, assignee=None, opened=None, why=None, project=None, compone
             did_change = True
             radar.sourceChanges = '\n'.join(source_changes)
 
+        if see_also:
+            sys.stderr.write('Radar does not support the see_also field at this time\n')
+            return None
```
I guess, kinda. I mean we could accept links to other Radar issues, but maybe that's Future Work.

gsnedders commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
```diff
@@ -748,3 +748,14 @@ def test_source_changes(self):
+            tracker = bugzilla.Tracker(self.URL)
+            self.assertEqual(tracker.issue(1).references, [])
+            self.assertTrue(tracker.issue(1).add_related_link('2'))
+            self.assertEqual(tracker.issue(1).references, [tracker.issue('2')])
```
I don't understand why this is intermixed with other `references`. Like sure, technically this is the case for the mock given the change in `Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py`, but I don't see how that would be true for the real Bugzilla?

briannafan replied:
`references` is a property that really only exists in `webkitbugspy` since it refers to any bugs that are "referenced" in the current one. After iterating through the comments, we actually append all issues from the `see_also` field to references. 

I would expect this test to accurately reflect "real" Bugzilla/webkitbugspy behaviour insofar as adding a bug to `see_also` would result it being populated in `references`: See [line 313, bugzilla.py](https://github.com/WebKit/WebKit/blob/016eaf8324d1f1bc84be4aed4d6f052b489ca4c5/Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py#L313).

gsnedders replied:
> I would expect this test to accurately reflect "real" Bugzilla/webkitbugspy behaviour insofar as adding a bug to `see_also` would result it being populated in `references`: See [line 313, bugzilla.py](https://github.com/WebKit/WebKit/blob/016eaf8324d1f1bc84be4aed4d6f052b489ca4c5/Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py#L313).

That I was _not_ expecting.

I think I'd prefer we had some separate field, `related_issues` or something?

But maybe I should defer to @JonWBedard there?

briannafan replied:
Based on our discussion, I added a separate field!

gsnedders reviewed (changes requested):
(I probably should've set the previous review as "Request changes" given I don't understand some of the choices as to why `see_also` and `keywords` are being treated differently. So see above for comments.)

briannafan committed:
[webkitbugspy] Add ability to set see_also field for bugs
rdar://151183507
https://bugs.webkit.org/show_bug.cgi?id=292893

Reviewed by Jonathan Bedard and Sam Sneddon.

Set the 'see_also' field in bugzilla by calling Issue.add_related_link.
This should be passed in as a list of string URLs.

'see_also' is stored in Issue.related_links as a new property since links that are not currently
tracked by webkitbugspy (GitHub, etc.) would not be populated in Issue.references.

* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.populate):
(Tracker.set):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py:
(Issue.__init__):
(Issue):
(Issue.related_links):
(Issue.add_related_links):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py:
(Bugzilla._issue):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker.populate):
(Tracker.set):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/github_unittest.py:
(test_source_changes):
(test_related_links):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/radar_unittest.py:

Canonical link: https://commits.webkit.org/295169@main
